File: ts/brute_force_text.ts
'use strict'
/**
 * Generates an array of all possible combinations of CHARACTERS from 1 digit to MAX_LENGTH digits.
 * @description Generates an array of all possible combinations of CHARACTERS from 1 digit to MAX_LENGTH digits.
 * Warning: A long maximum length (eg. 12) with a large number of unique characters (eg. 62) will take a long time to generate or may not complete without optimization. 
 */

/**
 * The maximum number of characters
 * @type {number}
 * @example MAX_LENGTH = 8 will create outputs from 1 (inclusive) to 8 (inclusive) characters.
 */
const MAX_LENGTH:number = 8;

/**
 * Unique characters
 * @type {String}
 */
const CHARACTERS:String = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ !#$%\&()*+,-./:;<=>?@[\]^_{|}~";

/**
 * Array of brute-force generated strings.
 * @type {Array<String>}
 */
var text:Array<String> = [];


/**
 * Returns all combinations of CHARACTERS at the given length
 * @param {int} length Length of output
 * @returns {int} Count of combinations at length, {x | x = CHARACTERS.length ^ length}
 */
function combinations(length:number = 1):number {
    return Math.pow(CHARACTERS.length, length);
}


/**
 * Returns all possible combinations of CHARACTERS at lengths 1 to length
 * @param {int} length Maximum length of output
 * @returns {int} Combinations of CHARACTERS at lengths 1 to length
 */
function combinationsStacked(length:number = 1):number {
    if (length > 1) {
        return combinations(length) + combinationsStacked(length - 1);
    }
    return combinations();
}


/**
 * String equivalent of decimal value.
 * @param {number} num Integer to be converted to string.
 * @returns {String} String equivalent of decimal input.
 * @example An input of 0 will return the first character from the selection of characters.
 * @see CHARACTERS
 */
function decimalToText(num:number):String {
    let str:String = "";
    while (num != 0) {
        str += CHARACTERS.charAt((num - 1) % CHARACTERS.length);
        num = Math.floor((num - 1) / CHARACTERS.length);
    }
    return str;
}

for (let i:number = 0; i <= combinationsStacked(MAX_LENGTH); i++) {
    text[i] = decimalToText(i);
}

console.log(text);

File: py/brute_force_text.py
"""
Brute force text generator - Character soup

Description: Generates an array of all possible combinations of CHARACTERS from 1 digit to MAX_LENGTH digits.
Warning: A long maximum length (eg. 12) with a large number of unique characters (eg. 62) will take a long time to generate or may not complete without optimization.
"""

import math as Math


MAX_LENGTH = 3
"""The maximum number of characters
"""

CHARACTERS = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ !#$%\&()*+,-./:;<=>?@[\]^_{|}~"
"""Characters to be used for brute force string generation
"""


def combinations(length=1):
    """Returns number of combinations

    Args:
        length (int, optional): Length of output. Defaults to 1.

    Returns:
        (int): Number of combinations of characters
    """
    return pow(len(CHARACTERS), length)


def combinations_factorial(length=1):
    """Returns number of combinations when combined. e.g. combinations(3) + combinations(2) + combinations(1)

    Args:
        length (int, optional): Maximum length of output, starting from 1. Defaults to 1.

    Returns:
        (int): Number of combinations of characters for lengths from 1 to length
    """
    if length > 1:
        return combinations(length) + combinations_factorial(length - 1)
    return combinations()


def decimalToText(num):
    """Converts a decimal number to a string using the CHARACTERS list

    Args:
        num (int): Number to count to in characters

    Returns:
        (str): Character, or characters, at given number
    """

    str = ""
    while num != 0:
        str += CHARACTERS[(num - 1) % len(CHARACTERS)]
        num = Math.floor((num - 1) / len(CHARACTERS))

    return str


def main():

    text = []  # List of all possible strings

    for i in range(combinations_factorial(MAX_LENGTH) + 1):
        text.insert(i, decimalToText(i))

    print(text)
    return


if __name__ == "__main__":
    main()
